gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/CallGraphBugChecker.java

```diff
@@ -0,0 +1,100 @@
+import com.google.common.graph.Traverser;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+public abstract class CallGraphBugChecker extends GradleGuideBugChecker {
+    /**
+     * A directed graph of "who is called by who" within this compilation unit (source file).
+     */
+    protected class MethodCallGraph {
+
+        /**
+         * Nodes are {@code MethodSymbol}s. There is an edge from {@code f} to {@code g} iff {@code f} is called by
+         * {@code g}.
+         */
+        @VisibleForTesting
+        protected final ImmutableGraph<MethodSymbol> callGraph;
+
+        public MethodCallGraph(VisitorState state) {
+            MutableGraph<MethodSymbol> mutableCallGraph =
+                    GraphBuilder.directed().allowsSelfLoops(true).build();
+
+            new SuppressibleTreePathScanner<Void, Optional<MethodSymbol>>(state) {
+                @Override
+                public Void visitMethod(MethodTree node, Optional<MethodSymbol> caller) {
+                    MethodSymbol methodSymbol = ASTHelpers.getSymbol(node);
+                    mutableCallGraph.addNode(methodSymbol);
+                    return super.visitMethod(node, Optional.of(methodSymbol));
+                }
+
+                @Override
+                public Void visitMethodInvocation(MethodInvocationTree node, Optional<MethodSymbol> caller) {
+                    MethodSymbol calleeSymbol = (MethodSymbol) ASTHelpers.getSymbol(node.getMethodSelect());
+                    mutableCallGraph.addNode(calleeSymbol);
+
+                    if (caller.isPresent()) {
+                        MethodSymbol callerSymbol = caller.get();
+                        mutableCallGraph.putEdge(calleeSymbol, callerSymbol);
+                    }
+
+                    return super.visitMethodInvocation(node, caller);
+                }
+            }.scan(state.getPath().getCompilationUnit(), Optional.empty());
+
+            callGraph = ImmutableGraph.copyOf(mutableCallGraph);
+        }
+
+        /**
+         * Returns the set of methods that call {@code target} transitively within this compilation unit (source file).
+         */
+        public Set<MethodSymbol> transitiveCallers(MethodSymbol target) {
+            Set<MethodSymbol> result = new HashSet<>();
+            for (MethodSymbol node : Traverser.forGraph(callGraph).depthFirstPreOrder(target)) {
+                result.addAll(callGraph.successors(node));
+            }
+            return result;
+        }
+    }
+
+    private MethodCallGraph callGraph;
+
+    /**
+     * The first call to this method builds the call graph on the state's compilation unit.
+     * Subsequent calls will return the cached call graph.
+     */
```

Comment:
Error Prone [keeps an instance of each `BugChecker` around for the entirety of the compilation stage](https://github.com/google/error-prone/blob/c59a1c9bdac1fc4f35e5f842ab537cb6c5d594e0/check_api/src/main/java/com/google/errorprone/scanner/ErrorProneScanner.java#L176), it doesn't make one per compilation unit. I think this means that this call graph will be created on the first compilation unit it sees then reused on later ones?

I think we need to use a `WeakMap<CompilationUnitTree, MethodCallGraph>` here so we can just store them all in a map, then when the compilation unit is finished/garbage collected, it automatically gets removed from the map to not blow up memory usage.